### ISSUE TITLE
[WARN] agent: Node info update blocked by ACLs

### ISSUE THREAD

**Issue by jkwysdom** [hashicorp/consul#4979]
I am trying to upgrade a consul server and am running into some issues. I have downloaded 1.3.0 (what I have been testing with and chef cookbooks are configured for so not able to switch to 1.4.0), left the cluster using `./consul leave` where `./consul` is the path to the consul binary. 

I want to be able to start the consul server via a script but right now when I start it via manually entering the command to start it using consul agent, I get the error message `[WARN] agent: Node info update blocked by ACLs` and I also notice that I am unable to register health checks even with the correct ACL token in my `conf.json` file. 

Command I am using to start consul:

/work-disk/consul/consul agent -server -bootstrap-expect=1 -data-dir=/work-disk/consul_data -config-file=/work-disk/consul/conf.json -bind=10.7.1.5 -enable-script-checks=true

Contents of conf.json
```
{
  "acl_datacenter": "consul-testing",
  "acl_default_policy":"deny",
  "acl_down_policy": "deny",
  "acl_master_token": "*<master token>*",
  "encrypt": "*<enc token>*==",
  "datacenter": "consul-testing"
}
```

What is causing this error message? 

I should note that my health checks will not register either and I am passing the master token as `X-Consul-Token` in my header for my API calls.

**Comment by jkwysdom** [comment]
Using the consul binary does not help either, I tried `./consul info -token=<master_acl_token>`

**Comment by rboyer** [comment]
When ACLs are enabled there are multiple tokens that may need to be configured for a given Consul Agent for different high level functions which are outlined here (for your version): https://www.consul.io/docs/guides/acl-legacy.html#configuring-acls

Of note would be the internal agent operations which are controlled with `acl_agent_token` not `acl_master_token` which you have exclusively configured above. `acl_master_token` is used just for bootstrapping the ACL system.

I think that if you were to configure the agent token with an appropriate policy things the warnings and failures you are seeing should go away.

**Comment by jkwysdom** [comment]
> Of note would be the internal agent operations which are controlled with `acl_agent_token` not `acl_master_token` which you have exclusively configured above. `acl_master_token` is used just for bootstrapping the ACL system.

Oh didn't know that, I guess a lot has changed with the ACL system since I last set consul up back when 0.7.1 was the most recent.

> I think that if you were to configure the agent token with an appropriate policy things the warnings and failures you are seeing should go away.

Do you know where I could find an example of a proper configuration or how I would generate/build one?

**Comment by jkwysdom** [comment]
@rboyer I am a bit confused now I think about this more, I have the exact same config working perfectly fine if I recreate a new cluster but I am only experiencing this issue when I have done an upgrade. If the config works for a new setup should it not also work after an upgrade or would there be something in place from a 0.7.1 => 1.3.0 upgrade that would require extra configuration?

**Comment by rboyer** [comment]
The bootstrapping guide (for your version) helps you incrementally configure and setup the relevant agent flags, configuration options, and ACL policies to get started: https://www.consul.io/docs/guides/acl-legacy.html#bootstrapping-acls

When you say you are experiencing this issue only when doing an upgrade, is this just for your 1-node test setup as described above or a more realistically sized cluster? Are you updating from 0.7.1 as above or something newer?

**Comment by jkwysdom** [comment]
@rboyer 
> The bootstrapping guide (for your version) helps you incrementally configure and setup the relevant agent flags, configuration options, and ACL policies to get started: https://www.consul.io/docs/guides/acl-legacy.html#bootstrapping-acls

Thanks will check it out and see what I can figure out.

> When you say you are experiencing this issue only when doing an upgrade, is this just for your 1-node test setup as described above or a more realistically sized cluster? Are you updating from 0.7.1 as above or something newer?

This is for a 2(sometimes 3) node cluster (how big our prod env is per customer), and yes updating from 0.7.1 to 1.3.0   I know best practice is a 3 or 5 node cluster but that would require us adding in more servers just for consul and we no chance I can convince anyone else to add a server just for that.

**Comment by rboyer** [comment]
Since you are jumping so many versions it would also be a good idea to review the upgrade notes for major things that have changed: https://www.consul.io/docs/upgrade-specific.html

**Comment by rboyer** [comment]
I think the main reason your node updates are failing at the agent level after the upgrade is to do with the version 0.8.0 upgrade that added more parts of consul under ACL control. See https://www.consul.io/docs/upgrade-specific.html#version-8-acls-are-now-opt-out.

If I bring up a 0.7.1 agent as above (sans the `-enable-script-checks` flag which did not exist) it works fine. If I upgrade to 0.8.0 the errors and warnings you saw show up because those endpoints are now being blocked by acl policy until an appropriate agent acl token is created and configured.

**Comment by rboyer** [comment]
Also if you are deploying 1.3.0 look to see if `enable-local-script-checks` works for your use case instead of `enable-script-checks`: https://www.consul.io/docs/agent/options.html#_enable_local_script_checks

If you are primarily having chef write out small configuration file snippets in the consul configuration directory to register services with script checks then `enable-local-script-checks` is more tailored specifically to that situation.

Also it is worth pointing out that v1.3.1 is out with a minor bugfix.

**Comment by jkwysdom** [comment]
@rboyer  Ah okay so I just need to create an `acl_agent_token` then? 

I register all of my checks via the API so I don't think the local checks flag would be effective in my case.

Also yes I saw that 1.3.1 was out and I checked the change log and it should not effect us as we do not use that part of the API. Our use case has us pretty much only using the check and service part of the API along with health state checking.

**Comment by rboyer** [comment]
You should be good to go if you create an ACL agent token with an appropriate policy and reconfigure your agents to use that ACL token.

**Comment by jkwysdom** [comment]
@rboyer Thanks, I have to rebuild my test ENV first and I will give that a try and see if it works.

**Comment by jkwysdom** [comment]
@rboyer I managed to set up a test ENV today and did the upgrade process and when I did it, it started consul no issues like before but I am getting a error 500 now when I run my chef cookbook which tries to reregister checks or get key value pairs.

I start consul with the command `/work-disk/consul/consul agent -server -data-dir /work-disk/consul_data -config-file /work-disk/consul/conf.json -bind 10.7.1.5 -enable-script-checks=true` 

Then I use the command `/work-disk/consul/consul join 10.7.1.6` to join back to the consul cluster that I had before (the other node) and get the below result from it.

```
Error joining address '10.7.1.6': Unexpected response code: 403 (Permission denied)
Failed to join any nodes.
```

The 10.7.1.6 node is still running the 0.7.1 as I understood from the docs that you need to run the join command again to rejoin the cluster before upgrading the other servers in the cluster

**Comment by jkwysdom** [comment]
@rboyer this seems to be an issue with either my consul implementation for 1.3 or an issue with 1.3 though I doubt that highly. I have now tried to set up another new instance of 1.3 and I am running into the exact same joining error and all of the same commands produce the same error as during the upgrade. 

Any insight into what might be happening?

Here is my config file for consul:

```
{
  "acl_datacenter": "consul-testing",
  "acl_default_policy":"deny",
  "acl_down_policy": "deny",
  "acl_master_token": "<token A>",
  "acl_token": "<token A>",
  "acl_agent_token":"<token B>",
  "encrypt": "<encrypt token>==",
  "datacenter": "consul-testing"
}
```

**Comment by ChipV223** [comment]
@jkwysdom : One thing you don't want to do is give `acl_master_token` and `acl_token` the same token. You should make a specific token for `acl_token` and have the appropriate policy.

Do you happen to have any debug logging that we can review?

**Comment by pearkes** [comment]
Given we haven't heard anything based on our suggestions/questions above I'm going to close this issue, but I encourage you to comment and we can re-open it if you want to pick this up again.

Alternatively, if things have changed dramatically, feel free to create a new issue or PR.

**Comment by Nmishin** [comment]
For anyone who will find this issue:
I have the same issue and resolve it by using [acl_agent_master_token](https://www.consul.io/docs/agent/options.html#acl_agent_master_token) because when consul cluster bootstrapped it can't validate any tokens include acl_master_token.

**Comment by nbari** [comment]
Just in case, the problem I had is that my agent policy is:

```
node_prefix "consul-" {
  policy = "write"
}
```

But my cluster nodes had a different `node_name` (Not starting with `consul-`) 

The acl sections I have it like:

```
acl {
  enabled = true
  default_policy = "deny"
  enable_token_persistence = true
  tokens {
    agent = "xxx-xxx-xxx..."
  }
}
```
